### PR TITLE
Organize the operation module documentation page

### PR DESCRIPTION
**Pull request by albi3ro**
The `operation` module defines a lot of things. As such, the automatically generated class summary is a bit large.

This PR breaks up the generated table of contents by the type object. Instead of all the classes being mixed together, there is a section for base classes, a section for errors, a section for Boolean functions, and a section for other.

This will force the contents of this module to be manually documented instead of automatically generated by sphinx, but I think it's worth the added organization.

It should be noted that the attributes are documented here, even though their code is placed in `ops/qubit/attributes`. 

In addition, `operation_derivative` could be moved to `ops/functions`. `WiresEnum`, `AllWires`, and `AnyWires` could be moved into the `wires` module.

### CODE REVIEW

**Comment by timmysilv** [comment]
thanks for starting this up, operation.py could definitely use some love. This change appears to do two things:
1. Re-arrange documentation to bring additional clarity as to what's in operation.py
2. Move the documentation definition from `doc/` to `pennylane/`

It's certainly great to have clearer documentation, but I don't love the requirement of manually doing it. `operation.py` changes at least once a week, and often more frequently than that. I don't think the state of that docs page is severe enough to add this expectation from contributors and reviewers alike. I see that the goal here is clearer documentation, but I think this points to a bigger problem which is that we should have clearer _code_. If you believe that `operation.py` is a special case that deserves this particular treatment, I could be convinced. At only a glance, I'd guess that we should instead re-arrange `operation.py` into an `operation/` module, and do any necessary docs touch-ups in the `doc/` folder. I rank the goal of "code is easy to maintain" higher than "documentation is clearer" in this particular scenario (because again, I think the docs are bearable today, but this is all personal opinion).

As for the definition move, is this necessary to add granularity? If it can be done in `doc/`, I'd rather leave it there.

tl;dr can we instead split `operation.py` into a module?

**Comment by trbromley** [comment]
I'm tagging @BM7878 for a review to keep him aware, though this particular change is mostly developer-facing documentation.

**Comment by josh146** [comment]
From a developer point of view, based on the location of this doc, it does feel like something that is as automatically generated as possible may be ideal? 
image

**Comment by albi3ro** [comment]
Given that `operation.py` is responsible for holding one of the core data structures for PennyLane, this file should only be changed by either:
1) bug fixes
2) planned improvements with widespread approval

Bug fixes should not change the contents. Planned improvements with widespread approval should have the documentation triply checked anyway.

The fact that this file does too much is a separate problem outside the scope of this quick PR.

**Review by BM7878 — APPROVED** [review]

**Review by timmysilv — COMMENTED** [review]
agreed on not making hasty changes to the file itself. so in other words, I'm open to talk about re-working this file. though i'm not sure I agree that documentation should be so close to code in this case, I'm fine with doing it.

My main hesitation that remains is the lack of automation. What's the expectation from reviewers/contributors? It wouldn't be the best if new features don't get documented because this step was forgotten.

**Comment by albi3ro** [comment]
@timmysilv There's an expectation for contributors and reviewers to check the documentation build and make sure any docstrings for new code is displaying correctly.  If the new object is not showing up, then the developer should go back and figure out why.

**Review by timmysilv — APPROVED** [review]
although the doc mechanism changes slightly here, that expectation is pretty clearly set for the codebase. docs lgtm, thank you! :shipit: :shipit:
